service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ObservationStatementMapper.java

```diff
@@ -75,23 +76,46 @@ public String mapObservationToObservationStatement(Observation observation, bool
             observation.getInterpretation().getCoding().stream()
                 .filter(this::isInterpretationCode)
                 .findFirst()
-                .ifPresent(coding ->
-                    observationStatementTemplateParametersBuilder.interpretation(
-                        structuredObservationValueMapper.mapInterpretation(coding))
-            );
+                .ifPresent(coding -> observationStatementTemplateParametersBuilder.interpretation(
+                    structuredObservationValueMapper.mapInterpretation(coding)
+                ));
         }
 
         if (observation.hasPerformer()) {
-            final String participantReference = idMapper.get(observation.getPerformerFirstRep());
-            final String participantBlock = participantMapper
-                .mapToParticipant(participantReference, ParticipantType.PERFORMER);
-            observationStatementTemplateParametersBuilder.participant(participantBlock);
+            observationStatementTemplateParametersBuilder.participant(buildParticipant(observation));
         }
 
         return TemplateUtils.fillTemplate(OBSERVATION_STATEMENT_EFFECTIVE_TIME_TEMPLATE,
             observationStatementTemplateParametersBuilder.build());
     }
 
+    private String buildParticipant(Observation observation) {
+        Optional<Reference> practitionerReference = getReferenceTo(observation, ResourceType.Practitioner);
+        Optional<Reference> organizationReference = getReferenceTo(observation, ResourceType.Organization);
+
+        String participantReference;
+        if (practitionerReference.isPresent() && organizationReference.isPresent()) {
+            participantReference = messageContext.getAgentDirectory().getAgentRef(practitionerReference.get(), organizationReference.get());
+        } else if (practitionerReference.isPresent()) {
+            participantReference = messageContext.getAgentDirectory().getAgentId(practitionerReference.get());
+        } else if (organizationReference.isPresent()) {
+            participantReference = messageContext.getAgentDirectory().getAgentId(organizationReference.get());
+        } else {
+            throw new EhrMapperException(
+                "Invalid performer reference in observation resource with id: " + observation.getIdElement().getIdPart()
+            );
+        }
```

Comment:
This might be a little easier to follow
```suggestion
        String participantReference;
        if (practitionerReference.isPresent() && organizationReference.isPresent()) {
            participantReference = messageContext.getAgentDirectory().getAgentRef(practitionerReference.get(), organizationReference.get());
        } else {
            var ref = practitionerReference.orElseGet(() -> organizationReference
                .orElseThrow(() -> new EhrMapperException("Invalid performer reference in observation resource with id: "
                    + observation.getIdElement().getIdPart())));
            participantReference = messageContext.getAgentDirectory().getAgentId(ref);
        }
```
Mostly I like it more because it doesn't use `Optional::get`, which always makes me nervous. What do you think?

Reply:
I think it's better thanks